### PR TITLE
Managed: Initial ManagedPoolController implementation

### PR DESCRIPTION
**Pull request by EndymionJkb**

### CODE REVIEW

File: pkg/pool-utils/contracts/controllers/ManagedPoolController.sol [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    function setSwapEnabled(bool swapEnabled) external virtual override onlyManager withBoundPool {
+        _require(canDisableSwaps(), Errors.UNAUTHORIZED_OPERATION);
+
+        IControlledManagedPool(pool).setSwapEnabled(swapEnabled);
+    }
+
+    /**
+     * @dev Pass a call to ManagedPool's setMustAllowlistLPs through to the underlying pool. This could
+     * be restricted in various ways. For instance, we could allow it to change state only once, or only
+     * in one direction, but there seems no compelling reason to do so in the reference controller.
```
Comment (TomAFrench) [inline pkg/pool-utils/contracts/controllers/ManagedPoolController.sol:166]:
```suggestion
     * in one direction, but there seems to be no compelling reason to do so in the reference controller.
```

File: pkg/pool-utils/contracts/controllers/ManagedPoolController.sol [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+            .insertBool(managedRights.canChangeWeights, _CHANGE_WEIGHTS_OFFSET)
+            .insertBool(managedRights.canDisableSwaps, _DISABLE_SWAPS_OFFSET)
+            .insertBool(managedRights.canSetMustAllowlistLPs, _MUST_ALLOWLIST_LPS_OFFSET);
+
+        // Last two separated to avoid "stack too deep"
+        return
+            permissions.insertBool(managedRights.canChangeTokens, _CHANGE_TOKENS_OFFSET).insertBool(
+                managedRights.canSetCircuitBreakers,
+                _CIRCUIT_BREAKERS_OFFSET
+            );
```
Comment (TomAFrench) [inline pkg/pool-utils/contracts/controllers/ManagedPoolController.sol:91]:
```suggestion
       bytes32 permissions = super.encodePermissions(baseRights);

        return
            permissions
                .insertBool(managedRights.canChangeWeights, _CHANGE_WEIGHTS_OFFSET)
                .insertBool(managedRights.canDisableSwaps, _DISABLE_SWAPS_OFFSET)
                .insertBool(managedRights.canSetMustAllowlistLPs, _MUST_ALLOWLIST_LPS_OFFSET)
                .insertBool(managedRights.canChangeTokens, _CHANGE_TOKENS_OFFSET)
                .insertBool(managedRights.canSetCircuitBreakers, _CIRCUIT_BREAKERS_OFFSET);
```

Comment (TomAFrench) [reply pkg/pool-utils/contracts/controllers/ManagedPoolController.sol:91]: This feels cleaner to me and gets around the stack too deep issues as well.

File: pkg/pool-utils/contracts/controllers/ManagedPoolController.sol [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    /**
+     * @dev Update weights linearly from the current values to the given end weights, between startTime
+     * and endTime.
+     */
+    function updateWeightsGradually(
+        uint256 startTime,
+        uint256 endTime,
+        uint256[] memory endWeights
```
Comment (TomAFrench) [inline pkg/pool-utils/contracts/controllers/ManagedPoolController.sol:143]:
```suggestion
        uint256[] calldata endWeights
```

Isn't this free gas savings?

File: pkg/pool-utils/contracts/controllers/ManagedPoolController.sol (hunk 2 above)
Comment (EndymionJkb) [reply pkg/pool-utils/contracts/controllers/ManagedPoolController.sol:91]: Yes; I agree that's much cleaner. Will do.

File: pkg/pool-utils/contracts/controllers/ManagedPoolController.sol [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    using WordCodec for bytes32;
+
+    // There are five managed pool rights: all corresponding to permissioned functions of ManagedPool.
+    struct ManagedPoolRights {
+        bool canChangeWeights;
+        bool canDisableSwaps;
+        bool canSetMustAllowlistLPs;
+        bool canSetCircuitBreakers;
+        bool canChangeTokens;
+    }
```
Comment (TomAFrench) [inline pkg/pool-utils/contracts/controllers/ManagedPoolController.sol:41]: Do we want to move these to the interface? Potentially not as `IControlledManagedPool` is generic to all `ManagedPool` controllers right?

**Review by TomAFrench — APPROVED** [review]
Looks good to me. Slight question on the best place to put the struct definition but will defer to your judgement.

File: pkg/pool-utils/contracts/controllers/ManagedPoolController.sol (hunk 4 above)
Comment (EndymionJkb) [reply pkg/pool-utils/contracts/controllers/ManagedPoolController.sol:41]: Yes, I'm inclined to leave them in the contracts, where the coupling is a lot tighter. You could theoretically have a right that isn't associated with a function.

We could also move the BasePoolRights - but the BasePoolController just has the initialize function.

Comment (TomAFrench) [reply pkg/pool-utils/contracts/controllers/ManagedPoolController.sol:41]: Cool, just wanted to float the topic.

**Pull request merged**